swhttpd.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
+import sys, logging
+import redis
+import syslog
+import socket
+import re
+import tempfile
+import SimpleHTTPServer
+import SocketServer
+import time
+
+import config
+
+def log(*args):
+  print time.strftime("%Y-%m-%d %H:%M:%S") + ':', ' '.join(args)
+  syslog.syslog(syslog.LOG_INFO, ' '.join(args))
+
+def error(*args):
+  print time.strftime("%Y-%m-%d %H:%M:%S") + ': ERROR:', ' '.join(args)
+  syslog.syslog(syslog.LOG_ERR, ' '.join(args))
+
+class swbootHttpHandler(SimpleHTTPServer.SimpleHTTPRequestHandler):
+  def do_GET(self):
+    db = redis.Redis()
+    switch = db.get(self.client_address[0])
+    model = db.get('client-{}'.format(self.client_address[0]))
+    if switch == None or model == None:
+      log("Switch not found:", self.client_address[0])
+      self.send_error(404, "File not found")
+      return None
+    if self.path == "/juniper-confg":
+      log("Generating Juniper config for",
+          self.client_address[0], "name =", switch)
+      f = tempfile.TemporaryFile()
+      f.write(config.generate(switch, model))
+      content_length = f.tell()
+      f.seek(0)
+
+      self.send_response(200)
+      self.send_header("Content-type", "application/octet-stream")
```

Comment:
I guess these are needed for Juniper to accept the config? Is there a source for this or just trial&error?

Reply:
We can't use the built in `send_head()` function since that assumes we're trying to send an actual file and tries to look up the file type and length of it. We generate the file on the fly and need to send each header individually. `Content-type` probably isn't needed, but doesn't hurt. There's nothing Juniper specific here.